File: app/src/main/java/com/timotiusoktorio/pencake/ui/addtocart/AddToCartViewModel.java
package com.timotiusoktorio.pencake.ui.addtocart;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.Nullable;

import com.timotiusoktorio.pencake.data.model.CartItem;
import com.timotiusoktorio.pencake.data.model.Product;
import com.timotiusoktorio.pencake.data.source.DataManager;

class AddToCartViewModel extends ViewModel {

    final MutableLiveData<Integer> quantityLiveData = new MutableLiveData<>();
    final MutableLiveData<Integer> subtotalLiveData = new MutableLiveData<>();
    final MutableLiveData<Boolean> exitFlagLiveData = new MutableLiveData<>();

    private DataManager dataManager;
    private Product product;
    private CartItem cartItem;
    private Integer selectedSizeIndex;

    AddToCartViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    void setInitialSizeAndQuantity(Product product, @Nullable CartItem cartItem) {
        this.product = product;
        this.cartItem = cartItem;

        if (cartItem == null) {
            setSelectedSizeIndex(0);
            setQuantity(1);
        } else {
            setSelectedSizeIndex(cartItem.getSelectedSizeIndex());
            setQuantity(cartItem.getQuantity());
        }
    }

    void setSelectedSizeIndex(Integer index) {
        selectedSizeIndex = index;
        setSubtotal();
    }

    void setQuantity(Integer quantity) {
        quantityLiveData.setValue(quantity);
        setSubtotal();
    }

    private void setSubtotal() {
        Integer quantity = quantityLiveData.getValue();
        if (quantity == null) {
            return;
        }
        Integer price = product.getPrices().get(selectedSizeIndex).getPrice();
        subtotalLiveData.setValue(price * quantity);
    }

    void onMinusButtonClick() {
        Integer quantity = quantityLiveData.getValue();
        if (quantity != null) {
            if (quantity > 1) {
                quantity--;
                setQuantity(quantity);
            }
        }
    }

    void onPlusButtonClick() {
        Integer quantity = quantityLiveData.getValue();
        if (quantity != null) {
            quantity++;
            setQuantity(quantity);
        }
    }

    void onSaveCartClick(String specialRequests) {
        if (cartItem == null) {
            CartItem newCartItem = new CartItem(null, product.toJson(), selectedSizeIndex, quantityLiveData.getValue(), subtotalLiveData.getValue(), specialRequests);
            dataManager.addCartItem(newCartItem);
        } else {
            cartItem.setSelectedSizeIndex(selectedSizeIndex);
            cartItem.setQuantity(quantityLiveData.getValue());
            cartItem.setSubtotal(subtotalLiveData.getValue());
            cartItem.setSpecialRequests(specialRequests);
            dataManager.updateCartItem(cartItem);
        }
        exitFlagLiveData.setValue(true);
    }

    void removeFromCart() {
        if (cartItem != null) {
            dataManager.deleteCartItem(cartItem);
            exitFlagLiveData.setValue(true);
        }
    }

    boolean isUpdatingCart() {
        return cartItem != null;
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/addtocart/AddToCartViewModel.kt
package com.timotiusoktorio.pencake.ui.addtocart

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.timotiusoktorio.pencake.data.model.CartItem
import com.timotiusoktorio.pencake.data.model.Product
import com.timotiusoktorio.pencake.data.source.DataManager

class AddToCartViewModel(private val dataManager: DataManager,
                         private val product: Product,
                         private val cartItem: CartItem?) : ViewModel() {

    val quantityLiveData = MutableLiveData<Int>()
    val subtotalLiveData = MutableLiveData<Int>()
    val exitFlagLiveData = MutableLiveData<Boolean>()

    private var selectedSizeIndex = 0
    private var specialRequests = ""

    init {
        if (cartItem == null) {
            setSelectedSizeIndex(0)
            setQuantity(1)
        } else {
            setSelectedSizeIndex(cartItem.selectedSizeIndex)
            setQuantity(cartItem.quantity)
        }
    }

    fun setSelectedSizeIndex(index: Int) {
        selectedSizeIndex = index
        calculateSubtotal()
    }

    fun setQuantity(quantity: Int) {
        quantityLiveData.value = quantity
        calculateSubtotal()
    }

    private fun calculateSubtotal() {
        quantityLiveData.value?.let {
            val price = product.prices[selectedSizeIndex].price
            subtotalLiveData.value = price * it
        }
    }

    fun setSpecialRequests(specialRequests: String) {
        this.specialRequests = specialRequests
    }

    fun decreaseQuantity() {
        quantityLiveData.value?.let { if (it > 1) setQuantity(it - 1) }
    }

    fun increaseQuantity() {
        quantityLiveData.value?.let { setQuantity(it + 1) }
    }

    fun addToCart() {
        if (cartItem == null) createCartItem()
        else updateCartItem(cartItem)
        exitFlagLiveData.value = true
    }

    private fun createCartItem() {
        val newCartItem = CartItem(
                productJson = product.toJson(),
                selectedSizeIndex = selectedSizeIndex,
                quantity = quantityLiveData.value ?: 0,
                subtotal = subtotalLiveData.value ?: 0,
                specialRequests = specialRequests
        )
        dataManager.addCartItem(newCartItem)
    }

    private fun updateCartItem(cartItem: CartItem) {
        cartItem.selectedSizeIndex = selectedSizeIndex
        cartItem.quantity = quantityLiveData.value ?: cartItem.quantity
        cartItem.subtotal = subtotalLiveData.value ?: cartItem.subtotal
        cartItem.specialRequests = specialRequests
        dataManager.updateCartItem(cartItem)
    }

    fun removeFromCart() {
        cartItem?.let {
            dataManager.deleteCartItem(it)
            exitFlagLiveData.value = true
        }
    }
}